### PR TITLE
Address performance issue in #550

### LINKED ISSUE #550: volatile array in device-side memset (performance issue)
With @jjennychen, we noticed that the device-side memset for CHIP-SPV was much slower than writing our own loop to do a memset. With @pengtu we narrowed this down to the fact that the array in memset (and memcpy) is declared as volatile (https://github.com/CHIP-SPV/chipStar/blob/96d21716110212530f663342c9ab214b9b866061/bitcode/devicelib.cl#L446).

Is there a reason this is volatile? I can't think of why we would need that for the memset, but I could be missing something. When we tested without the volatile, we saw much better performance.

Thanks!

### PR DESCRIPTION
**Pull request by linehill**
... by removing volatile keywords in __chip_mem{set,cpy}.

Closes #550.

### CODE REVIEW

**Review by pjaaskel — APPROVED** [review]

File: llvm_passes/HipLowerMemset.cpp [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// Expand llvm.memset intrinsics to loops before handing the bitcode to
+// llvm-spirv translator. The translator expands memset intrinsics to emulation
+// function with a name encoding the destination pointer type
+// (e.g. "spirv.spirv.llvm_memset_p0_i32..."). During reverse translation,
+// llvm-spirv attempts to recreate the original memset intrinsic which can be
+// incorrect if we are jumping from opaque pointer world to typed pointer world
+// or vice versa. LLVM module verification may fail with the following message:
```
Comment (pjaaskel) [inline llvm_passes/HipLowerMemset.cpp:15]: Is this a SPIR-V translator bug? Can you add a link to the issue here?

File: llvm_passes/HipLowerMemset.cpp [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+//   type arguments!
+//
+// llvm.memcopy does not have this issue as it gets translated to actual SPIR-V
+// instruction (OpCopyMemorySized).
+//
+// (c) 2023 Henry Linjamäki / Intel
+//===----------------------------------------------------------------------===//
+#include "HipLowerMemset.h"
+
+#include "llvm/ADT/SmallPtrSet.h"
```
Comment (pjaaskel) [inline llvm_passes/HipLowerMemset.cpp:27]: Use "system include" (#include <>) if not referring to headers in our own tree.

File: llvm_passes/HipLowerMemset.cpp [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+#include "llvm/IR/Function.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/Passes/PassBuilder.h"
+#include "llvm/Passes/PassPlugin.h"
+#include "llvm/Transforms/Utils/LowerMemIntrinsics.h"
+
+using namespace llvm;
+
+static bool lowerMemsets(Function &F) {
```
Comment (pjaaskel) [inline llvm_passes/HipLowerMemset.cpp:37]: Are you sure there is not such a utility function in upstream LLVM? Would make sense.

**Review by pjaaskel — APPROVED** [review]
If there is no SPIR-V translator issue, pls open one. Otherwise LGTM.

File: llvm_passes/HipLowerMemset.cpp (hunk 1 above)
Comment (linehill) [reply llvm_passes/HipLowerMemset.cpp:15]: I'll file the issue later.

File: llvm_passes/HipLowerMemset.cpp (hunk 3 above)
Comment (linehill) [reply llvm_passes/HipLowerMemset.cpp:37]: expandMemSetAsLoop() is an utility function provided by the upstream LLVM.

**Comment by linehill** [comment]
> If there is no SPIR-V translator issue, pls open one.

Opened one and added a link to it in the HipLowerMemset.cpp.

**Comment by pvelesko** [comment]
@pjaaskel 
```
1299: Test command: /home/pvelesko/chipStar/addLibCeed/build/tests/runtime/TestMemFunctions
1299: Working Directory: /home/pvelesko/chipStar/addLibCeed/build/tests/runtime
1299: Test timeout computed to be: 10000000
1299: Fails to verify module: Intrinsic name not mangled correctly for type arguments! Should be: llvm.memset.p1i8.i64
1299: void (i8 addrspace(1)*, i8, i64, i1)* @llvm.memset.p1.i64
1299: CHIP error [TID 517151] [1692337746.876520025] : Program BUILD LOG for device #0:cpu-12th Gen Intel(R) Core(TM) i3-12100:
1299: Device cpu-12th Gen Intel(R) Core(TM) i3-12100 failed to build the program
1299: 
1299: 
1299: CHIP error [TID 517151] [1692337746.876553120] : hipErrorNotInitialized (CL_BUILD_PROGRAM_FAILURE ) in /home/pvelesko/chipStar/addLibCeed/src/backend/OpenCL/CHIPBackendOpenCL.cc:729:compile
1299: 
1299: CHIP error [TID 517151] [1692337746.876616195] : Caught Error: hipErrorNotInitialized
1/1 Test #1299: TestMemFunctions .................***Failed    0.55 sec
```

**Comment by pvelesko** [comment]
```
/opt/install/llvm/15.0/bin/opt: symbol lookup error: /opt/actions-runner/_work/chipStar/chipStar/build/lib/libLLVMHipSpvPasses.so: undefined symbol: _ZN4llvm18expandMemSetAsLoopEPNS_10MemSetInstE
```

**Review by pvelesko — CHANGES_REQUESTED** [review]
build errors

**Comment by pvelesko** [comment]
This is specific to this PR

**Comment by pvelesko** [comment]
Fails for both GitHub runner via Brice's CI and on cupcake.

**Comment by pjaaskel** [comment]
```
nm -DC ~/local/lib/libLLVM-16.0.6.so | grep expandMemSetAsLoop
00000000017b9c50 T llvm::expandMemSetAsLoop(llvm::MemSetInst*)@@LLVM_16
```
I have that symbol. What about your LLVM so, @pvelesko ?

**Comment by linehill** [comment]
Also, what does `ldd /opt/install/llvm/15.0/bin/opt` say? Does the `opt` link to the LLVM it is supposed to?

**Comment by pvelesko** [comment]
> Also, what does `ldd /opt/install/llvm/15.0/bin/opt` say? Does the `opt` link to the LLVM it is supposed to?

```
╭─pvelesko@cupcake ~
╰─$ ldd /opt/install/llvm/15.0/bin/opt                                                                                                                                                                                                         130 ↵
	linux-vdso.so.1 (0x00007ffe8ade2000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f52a14b9000)
	libzstd.so.1 => /lib/x86_64-linux-gnu/libzstd.so.1 (0x00007f52a13ea000)
	libtinfo.so.6 => /lib/x86_64-linux-gnu/libtinfo.so.6 (0x00007f52a13b8000)
	libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f52a118c000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f52a10a5000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f52a1083000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f52a0e5b000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f52a470d000)
```

**Comment by pvelesko** [comment]
I don't have this library in the install dir? I'm using `scripts/configure_llvm.sh` to build and install llvm into `/opt/install/llvm/16.0` 

I have a couple of these older .so in my system paths but these are old and not used.
```
╭─pvelesko@cupcake /opt/install/llvm
╰─$ find / -name "libLLVM-*" 2> /dev/null
/usr/lib/llvm-15/lib/python3/dist-packages/lldb/libLLVM-15.so.1
/usr/lib/llvm-15/lib/python3/dist-packages/lldb/libLLVM-15.0.7.so.1
/usr/lib/llvm-15/lib/libLLVM-15.so.1
/usr/lib/llvm-15/lib/libLLVM-15.0.7.so.1
/usr/lib/llvm-15/lib/libLLVM-15.so
/usr/lib/llvm-15/lib/libLLVM-15.0.7.so
/usr/lib/llvm-14/lib/libLLVM-14.so.1
/usr/lib/llvm-14/lib/libLLVM-14.0.0.so
/usr/lib/llvm-14/lib/libLLVM-14.so
/usr/lib/llvm-14/lib/libLLVM-14.0.0.so.1
/usr/lib/x86_64-linux-gnu/libLLVM-15.so.1
/usr/lib/x86_64-linux-gnu/libLLVM-15.0.7.so.1
/usr/lib/x86_64-linux-gnu/libLLVM-15.so
/usr/lib/x86_64-linux-gnu/libLLVM-14.so.1
/usr/lib/x86_64-linux-gnu/libLLVM-13.so.1
/usr/lib/x86_64-linux-gnu/libLLVM-14.so
/usr/lib/x86_64-linux-gnu/libLLVM-13.so
/usr/lib/x86_64-linux-gnu/libLLVM-14.0.0.so.1
```

**Comment by pjaaskel** [comment]
I think this is a problem with the static linkage, which is an annoying problem with LLVM and the dynamically loaded passes: opt hasn't needed that symbol, thus hasn't linked it in and since it doesn't have a dynlib to load it from, it fails. If we link in that symbol statically to our pass then we might get multiple defs of that symbol if loading to a dynamically linked opt which also gets that symbol from the LLVM dynlib.

**Comment by linehill** [comment]
It could also be a fault in libLLVMHipSpvPasses.so as we currently don't link it to LLVM (dynamically or statically):

```
/mnt/md1/linehill/ws-chipstar-2/install/bin/clang++ -fPIC  -Wno-duplicate-decl-specifier -Wno-tautological-constant-compare  -Wno-c++20-extensions -Wno-unused-result -Wno-delete-abstract-non-virtual-dtor -Wno-deprecated-declarations -Wunused-command-line-argument -gdwarf-4 -g -shared  -o ../lib/libLLVMHipSpvPasses.so CMakeFiles/LLVMHipPasses.dir/HipPasses.cpp.o CMakeFiles/LLVMHipPasses.dir/HipDynMem.cpp.o CMakeFiles/LLVMHipPasses.dir/HipStripUsedIntrinsics.cpp.o CMakeFiles/LLVMHipPasses.dir/HipDefrost.cpp.o CMakeFiles/LLVMHipPasses.dir/HipPrintf.cpp.o CMakeFiles/LLVMHipPasses.dir/HipGlobalVariables.cpp.o CMakeFiles/LLVMHipPasses.dir/HipTextureLowering.cpp.o CMakeFiles/LLVMHipPasses.dir/HipAbort.cpp.o CMakeFiles/LLVMHipPasses.dir/HipEmitLoweredNames.cpp.o CMakeFiles/LLVMHipPasses.dir/HipWarps.cpp.o CMakeFiles/LLVMHipPasses.dir/HipKernelArgSpiller.cpp.o CMakeFiles/LLVMHipPasses.dir/HipLowerZeroLengthArrays.cpp.o CMakeFiles/LLVMHipPasses.dir/HipSanityChecks.cpp.o CMakeFiles/LLVMHipPasses.dir/HipLowerSwitch.cpp.o 
```
^ No linking to LLVM libraries here.

```
$ readelf -d lib/libLLVMHipSpvPasses.so | grep NEEDED
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libm.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
```

**Comment by pjaaskel** [comment]
> What is being statically linked right now?

Your 'opt', if I read your ldd output right.

@linehill yep, we need to link, but the static linkage might have the problem I described. It should have a solution though as it should be normal to link in out-of-tree LLVM passes as plugins (I recall struggling with that long time ago).

**Comment by pvelesko** [comment]
PoCL failrues when trying to use dynamic LLVM? @pjaaskel 
Maybe PoCL didn't get rebuilt once I switched LLVM? 
```
/home/runner/opt/llvm/16/bin/opt: symbol lookup error: /home/runner/work/chipStar/chipStar/build/lib/libLLVMHipSpvPasses.so: undefined symbol: _ZN4llvm18expandMemSetAsLoopEPNS_10MemSetInstE
```

**Comment by pvelesko** [comment]
@pjaaskel I forced the rebuild of PoCL and still:
```
[ 64%] Building CXX object samples/abort/CMakeFiles/abort.dir/abort.cc.o
/home/runner/opt/llvm/16/bin/opt: symbol lookup error: /home/runner/work/chipStar/chipStar/build/lib/libLLVMHipSpvPasses.so: undefined symbol: _ZN4llvm18expandMemSetAsLoopEPNS_10MemSetInstE

```

**Comment by linehill** [comment]
The cause of the issue is probably still the same as before. The plugin is loaded by statically built opt tool which does not provide symbols needed by the plugin. Made a patch (#645) for the plugin  so it links to the LLVM dynamically. Try if it fixes the problem (make sure you build the chipStar from scratch).

**Comment by pvelesko** [comment]
Does building llvm as dynamic not link opt dynamic as well?

**Comment by linehill** [comment]
> Does building llvm as dynamic not link opt dynamic as well?

I don't know LLVM build system well enough to answer the question. Like, is there a configuration that let one build shared LLVM library but still have binary tools statically linked? 

The recent lookup error suggests that the opt tool is statically linked but that should not be a matter anymore. I pushed a change that makes HipMemsetLowering work with statically linked opt tool.

**Comment by franz** [comment]
> Like, is there a configuration that let one build shared LLVM library but still have binary tools statically linked?

I think so. Using `-DLLVM_BUILD_LLVM_DYLIB=ON` together with `-DLLVM_LINK_LLVM_DYLIB=OFF`. The manual mentions: `Tools will only be linked to the libLLVM shared library if LLVM_LINK_LLVM_DYLIB is also ON.`

**Pull request merged**